# HW 2 
# Sai Raina

In [2]:
import pandas as pd
import numpy as np
from sklearn.metrics import f1_score

In [3]:
df = pd.read_excel(r'Grade.xlsx')

In [4]:
df.drop(['Unnamed: 4','Unnamed: 5'],axis=1,inplace=True)

In [5]:
df.head()

HW  Midterm  Final Pass/Fail
0  80       72     62         f
1  63       77     67         f
2  77       83     68         f
3  91       73     78         f
4  92       86     70         f

In [6]:
df['Pass/Fail'].value_counts()

p    18
f     9
Name: Pass/Fail, dtype: int64

In [7]:
X = df.iloc[:,:-1]
Y = df.iloc[:,-1]

In [8]:
from sklearn.linear_model import LogisticRegression

In [9]:
clf = LogisticRegression(random_state=0).fit(X, Y)

In [10]:
clf.score(X,Y)

0.9629629629629629

In [11]:
pred = clf.predict(X)

f1_score(Y,pred,pos_label='p')

0.9714285714285714

In [12]:
X = df.loc[:,["Final","HW"]]

In [13]:
# Final and HomeWork
clf1 = LogisticRegression(random_state=0).fit(X, Y)
print("Accuracy :",clf1.score(X,Y))

pred = clf1.predict(X)

print("F1 Score :",f1_score(Y,pred,pos_label='p'))

Accuracy : 0.8148148148148148
F1 Score : 0.8648648648648649


In [14]:
X = df.loc[:,["Final","Midterm"]]

In [15]:
# Midterm and Final
clf2 = LogisticRegression(random_state=0).fit(X, Y)
print("Accuracy :",clf2.score(X,Y))

pred = clf2.predict(X)

print("F1 Score :",f1_score(Y,pred,pos_label='p'))

Accuracy : 0.9259259259259259
F1 Score : 0.9444444444444444


In [16]:
X = df.loc[:,["HW","Midterm"]]

In [17]:
# HW and Midterm
clf3 = LogisticRegression(random_state=0).fit(X, Y)
clf3.score(X,Y)

0.8888888888888888

In [18]:
pred = clf3.predict(X)

f1_score(Y,pred,pos_label='p')

0.923076923076923

In [19]:
X = df.iloc[:,:-1]

In [20]:
from sklearn.ensemble import RandomForestClassifier
 
forest = RandomForestClassifier(n_estimators=500,
                                random_state=1)

forest.fit(X, Y.values.ravel())
importances = forest.feature_importances_

ft_imp = pd.DataFrame({'Feature Name':X.columns,'Importance':importances})
ft_imp

Feature Name  Importance
0           HW    0.329293
1      Midterm    0.324590
2        Final    0.346117

In [21]:
X = df.loc[:,["Final","Midterm"]]

In [34]:
from sklearn.metrics import confusion_matrix, classification_report

rf_clf = RandomForestClassifier(n_estimators=50,random_state=1,ccp_alpha=0.1)
rf_clf.fit(X,Y)

pred = rf_clf.predict(X)
acc = rf_clf.score(X,Y)
acc

0.9629629629629629

In [35]:
confusion_matrix(Y, pred)

array([[ 8,  1],
       [ 0, 18]], dtype=int64)

In [36]:
f1_score(Y,pred,pos_label='p')

0.972972972972973

In [25]:
print(classification_report(Y, pred))

              precision    recall  f1-score   support

           f       1.00      0.78      0.88         9
           p       0.90      1.00      0.95        18

    accuracy                           0.93        27
   macro avg       0.95      0.89      0.91        27
weighted avg       0.93      0.93      0.92        27



# The best model is the one that uses the features Midterm and Final scores to predict if the student will pass or fail.

## This is based on the fact that it achieves the highest accuracy. To combat the slight imabalance in the dataset we can look at the f1-score which takes into consideration both precision and recall. The model which uses these features also has the highest f1-score.

## I have used two models logistic regression and random forest classifier. Both perform similarly infact since the dataset is simple and small logistic regression does better and does not overfit. While random forest is slightly tending towards overfitting.

### Logistic Regression --> 
    Accuracy : 0.9259259259259259
    F1 Score : 0.9444444444444444
    
### Random Forest --> 
    Accuracy : 0.9629629629629629
    F1 Score : 0.972972972972973